### PR TITLE
LED

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Adds LongformerEncoderDecoder (LED) from @ibeltagy - see: https://github.com/allenai/longformer#longformer

Todo:

- [x] **Important**: position embeddings have to be cut to correctly convert original Bart-ilke checkpoints to LED. The reason is that Bart uses a position embedding hack because of which the embedding idx 0 and 1 are never used resulting in an embedding matrix that has a length of 1026 instead of 1024, see: https://github.com/huggingface/transformers/blob/88ef8893cd649cc2b4adb9885aba88c750118cff/src/transformers/models/bart/modeling_bart.py#L131. All LED checkpoints are hence cut to remove this hack in LED:
```python
model = LEDForConditionalGeneration.from_pretrained("./led-base-16384")
model.model.encoder.embed_positions.weight = torch.nn.Parameter(model.model.encoder.embed_positions.weight[2:, :])
model.model.decoder.embed_positions.weight = torch.nn.Parameter(model.model.decoder.embed_positions.weight[2:, :])
model.save_pretrained("./led-base-16384")
```

- [x] Make Pytorch integration tests pass. See `LEDIntegrationTests` in `tests/test_modeling_led.py`.
- [x] Add gradient_checkpointing
- [x] Make common tests work
- [x] Add convenient padding function so that input can be of whatever size and add global_attn logic to mask
- [x] Automatically create attention_mask in encoder if not provided
- [x] Finish PT version
- [x] Make TF version work
- [x] Add tips in docs for LED
- [x] Eval notebook: https://colab.research.google.com/drive/12INTTR6n64TzS4RrXZxMSXfrOd9Xzamo?usp=sharing
- [x] Nice to have: Fine-tune notebook: https://colab.research.google.com/drive/12LjJazBl7Gam0XBPy_y0CTOJZeZ34c2v?usp=sharing
- [ ] Make nice model cards

## TODO after PR is merged:
- [ ] Correctly add `# Copied from ....` statements from Bart and Longformer (this probably requires the Bart refactor to be merged before)
- [ ] Open issue regarding problems with TF save_model test
- [ ] Correct templates: delete unnecessary test for tf bart; add gradient checkpointing by default in PT

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/led/modeling_led.py
Comment (patrickvonplaten) [inline src/transformers/models/led/modeling_led.py:2209]: cc @patil-suraj @ibeltagy  -> we should probalby apply this to all seq2seq models. The saved cross-attention cache never needs to be re-ordered since it's always equal for all `beam_idx`. Since re-ordering relies on `index_select` which creates a new tensor in PT this re-ordering can actually consume a lot of memory. In the case of `Led` this saved ~2GB of RAM because the encoder_hidden_states can be **very** large

File: src/transformers/models/led/modeling_led.py
Comment (patrickvonplaten) [inline src/transformers/models/led/modeling_led.py:170]: @ibeltagy this line ensures that the whole class **has to be** exactly the same as the corresponding class in `LongformerSelfAttention` or else the tests would throw an error. This is our safety check to make sure that the original class (which in other libs would just be imported and re-used here) cannot change without this class being changed accordingly as well.

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:236]: `use_cache` should never be enabled for gradient_checkpointing

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (patrickvonplaten) [inline src/transformers/models/gpt2/modeling_gpt2.py:718]: gradient checkpointing was previously only possible with `use_cache=True` which does not make much sense -> change layer accordingly here.

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:236]:
```suggestion
```

File: tests/test_modeling_tf_led.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_led.py:285]: This is due to a more general bug in TF save IMO. Should be solved differently. See: https://github.com/huggingface/transformers/issues/9313

Comment (jplu) [reply tests/test_modeling_tf_led.py:285]: this test must be fixed before merging this PR. It means the work on TF Serving must be merged before to integrate the fix and then rebase to integrate all the fix 👍

Comment (patrickvonplaten) [reply tests/test_modeling_tf_led.py:285]: Actually, I think it's better if we merge this PR and afterward do the required changes (which I'm happy to do). We will have to merge 10+ currently open PRs + change TF Bart & all TF Bart-like models & add a new TF serving test & change the TF templates to finish TF serving. Making this model compliant with the "new" not-yet implemented TF serving logic is blocking this PR too much IMO.

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:24]: Please, use the `get_tf_activation` instead of `ACT2FN`.

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:1808]:
```suggestion
            shape_list(inputs["past_key_values"][0][0])[2] if inputs["past_key_values"] is not None else 0
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:2068]:
```suggestion
            num_tokens_to_copy = min(shape_list(self.final_logits_bias)[0], new_num_tokens)
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:453]:
```suggestion
        padding = tf.convert_to_tensor(
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:518]:
```suggestion
        paddings = tf.convert_to_tensor([[0, 0], [window_overlap, window_overlap], [0, 0]], dtype=tf.dtypes.int32)
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:578]:
```suggestion
        paddings = tf.convert_to_tensor([[0, 0], [0, 0], [0, 0], [0, window_overlap + 1]])
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [inline src/transformers/models/led/modeling_tf_led.py:125]: Is-it possible to start to use the `tf.keras.layers.experimental.EinsumDense` and remove the usage of the `Dense` layers+transpose as it is done in the BERT implementation? Or it will requires too much work?

**Review by jplu — APPROVED** [review]
Awesome work!!! Just left few tiny comments.

File: tests/test_modeling_tf_led.py
Comment (jplu) [reply tests/test_modeling_tf_led.py:285]: Iindeed it is blocking a bit too much for a later compliancy.

Just checked deeper to the implementation and the new coming template might solve this issue. Ok to merge then 👌

File: src/transformers/models/led/modeling_tf_led.py
Comment (patrickvonplaten) [reply src/transformers/models/led/modeling_tf_led.py:125]: Hmm...this class is just a copy of TFLongformer's self-attention. So I'd like to address this in a future PR as well. As soon as we fix TFLongformer's self-attention to have `tf.keras.layers.experimental.EinsumDense`, this code will automatically be overwritten.

File: src/transformers/models/led/modeling_tf_led.py
Comment (patrickvonplaten) [reply src/transformers/models/led/modeling_tf_led.py:578]: Good for me! Out of curiosity, why is this function preferred?

File: src/transformers/models/led/modeling_tf_led.py
Comment (patrickvonplaten) [reply src/transformers/models/led/modeling_tf_led.py:2068]: great catch! I'll correct this in the templates as well

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [reply src/transformers/models/led/modeling_tf_led.py:125]: Ok for me!

File: src/transformers/models/led/modeling_tf_led.py
Comment (jplu) [reply src/transformers/models/led/modeling_tf_led.py:578]: Oops my bad I should have left an explanation. This is because `tf.constant` doesn't work in graph mode when one of the values is a tensor (here `window_overlap + 1`)

File: tests/test_modeling_tf_led.py
Comment (LysandreJik) [inline tests/test_modeling_tf_led.py:2]:
```suggestion
# Copyright 2021 Iz Beltagy, Matthew E. Peters, Arman Cohan and The HuggingFace Inc. team. All rights reserved.
```

**Review by LysandreJik — APPROVED** [review]
Great job implementing this! Happy to hear the model templates made your life easier, too.

Left a few nits, but LGTM! 

PS: That 16384 embedding size is incredible!

File: src/transformers/models/longformer/modeling_tf_longformer.py
Comment (patrickvonplaten) [reply src/transformers/models/longformer/modeling_tf_longformer.py:799]: Yeah, apparantly `tf.convert_to_tensor` is safer here @jplu

File: src/transformers/models/led/modeling_led.py
Comment (patrickvonplaten) [reply src/transformers/models/led/modeling_led.py:2102]: Yeah, I'll do it!

File: src/transformers/models/led/modeling_led.py
Comment (sgugger) [inline src/transformers/models/led/modeling_led.py:1427]:
```suggestion
        decoder_attention_mask (:obj:`torch.LongTensor` of shape :obj:`(batch_size, target_sequence_length)`, `optional`):
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (sgugger) [inline src/transformers/models/led/modeling_tf_led.py:1420]:
```suggestion
        decoder_attention_mask (:obj:`tf.Tensor` of shape :obj:`(batch_size, target_sequence_length)`, `optional`):
```

**Review by sgugger — APPROVED** [review]
Looks great to me! Thanks for adding this new model!

File: src/transformers/models/led/modeling_led.py
Comment (sgugger) [inline src/transformers/models/led/modeling_led.py:1427]: Woops, `length` ends up being here twice. Sorry about that!

File: src/transformers/models/led/modeling_tf_led.py
Comment (sgugger) [inline src/transformers/models/led/modeling_tf_led.py:1420]: Same here.

File: src/transformers/models/led/modeling_led.py
Comment (patrickvonplaten) [reply src/transformers/models/led/modeling_led.py:1427]: Thanks for re-checking!

File: src/transformers/models/led/modeling_led.py
Comment (patrickvonplaten) [inline src/transformers/models/led/modeling_led.py:1427]:
```suggestion
        decoder_attention_mask (:obj:`torch.LongTensor` of shape :obj:`(batch_size, target_sequence_length)`, `optional`):
```

File: src/transformers/models/led/modeling_tf_led.py
Comment (patrickvonplaten) [inline src/transformers/models/led/modeling_tf_led.py:1420]:
```suggestion
        decoder_attention_mask (:obj:`tf.Tensor` of shape :obj:`(batch_size, target_sequence_length)`, `optional`):
```

**Comment by jplu** [comment]
@patrickvonplaten when you have time, can you fix the conflicts and apply the same updates merged in Longformer to LED. Thanks!

**Pull request merged**
